format/aggregate.go

```diff
@@ -0,0 +1,156 @@
+//
+// Aggregate is a formatter which can group up further formatter.
+// The `ApplyTo` settings will be pass on and overwritten in the child formatter.
+// This plugin could be useful to setup complex configs with metadata handling in more readable format.
+//
+// Parameters
+//
+// - ApplyTo: This value chooses the part of the message the formatting
+// should be applied to. Use "" to target the message payload; other values
+// specify the name of a metadata field to target.
+// This value will also used for further child modulators!
+// By default this parameter is set to "".
+//
+// - Modulators: Defines a list of child modulators to be applied to a message when
+// it arrives at this formatter. If a modulator changes the stream of a message
+// the message is NOT routed to this stream anymore.
+// By default this parameter is set to an empty list.
+//
+// Examples
+//
+// This example show a useful case for format.Aggregate plugin:
+//
+//  exampleConsumerA:
+//    Type: consumer.Console
+//    Streams: "foo"
+//    Modulators:
+//      - format.MetadataCopy:
+//          CopyToKeys: ["foo", "bar"]
+//      - format.Aggregate:
+//          ApplyTo: foo
+//          Modulators:
+//            - format.Base64Encode
+//            - format.Double
+//            - format.Envelope:
+//                Postfix: "\n"
+//      - format.Envelope:
+//          Postfix: "\n"
+//          ApplyTo: bar
+//
+//  # same config as
+//  exampleConsumerB:
+//    Type: consumer.Console
+//    Streams: "bar"
+//    Modulators:
+//      - format.MetadataCopy:
+//          CopyToKeys: ["foo", "bar"]
+//      - format.Base64Encode:
+//          ApplyTo: foo
+//      - format.Double:
+//          ApplyTo: foo
+//      - format.Envelope:
+//          Postfix: "\n"
+//          ApplyTo: foo
+//      - format.Envelope:
+//          Postfix: "\n"
+//          ApplyTo: bar
+//
+type Aggregate struct {
+	core.SimpleFormatter `gollumdoc:"embed_type"`
+	modulators           core.ModulatorArray
+}
+
+func init() {
+	core.TypeRegistry.Register(Aggregate{})
+}
+
+// Configure initializes this formatter with values from a plugin config.
+func (format *Aggregate) Configure(conf core.PluginConfigReader) {
+	applyTo := conf.GetString("ApplyTo", "")
+
+	// init modulator array
+	modulatorSettings := format.getModulatorSettings(applyTo, conf)
+
+	config := core.NewPluginConfig("", "format.Aggregate.Modulators")
+	config.Override("Modulators", modulatorSettings)
+	modulatorReader := core.NewPluginConfigReaderWithError(&config)
+
+	modulatorArray, err := modulatorReader.GetModulatorArray("Modulators", format.Logger, core.ModulatorArray{})
+	if err != nil {
+		conf.Errors.Push(err)
+	}
+
+	format.modulators = modulatorArray
+}
+
+// ApplyFormatter execute the formatter
+func (format *Aggregate) ApplyFormatter(msg *core.Message) error {
+	for _, modulator := range format.modulators {
+		modulateResult := modulator.Modulate(msg)
+		if modulateResult != core.ModulateResultContinue {
```

Comment:
discard should produce an error to be propagated to the wrapping modulate call (see formattermodulator.go)